### PR TITLE
Override connection for subscriptions

### PR DESCRIPTION
**Pull request by SzymonPobiega**
Related to https://github.com/Particular/NServiceBus.Persistence.Sql/issues/276

[Here](https://gist.github.com/SzymonPobiega/49e3d1b340755dee8edbfb2dc93725c1) the code to smoke test this feature as it is super hard to test it automatically (requires separate DBs etc.)

### CODE REVIEW

File: src/SqlPersistence/Config/SqlPersistenceConfig_Connection.cs [hunk 1]
```diff
@@ -19,14 +19,24 @@ public static void ConnectionBuilder(this PersistenceExtensions<SqlPersistence>
...
-            if (settings.TryGet("SqlPersistence.ConnectionBuilder", out Func<DbConnection> value))
+            if (settings.TryGet($"SqlPersistence.ConnectionBuilder.{storageType.Name}", out Func<DbConnection> value))
+            {
+                return value;
+            }
+            if (settings.TryGet("SqlPersistence.ConnectionBuilder", out value))
             {
                 return value;
             }
             throw new Exception("ConnectionBuilder must be defined.");
```
Comment (DavidBoike) [inline src/SqlPersistence/Config/SqlPersistenceConfig_Connection.cs]: I know this wasn't changed but maybe we could provide a better exception here, and also include the storage type it's trying to find.

File: src/SqlPersistence/ScriptRunner.cs [hunk 2]
```diff
@@ -29,30 +40,37 @@ public static class ScriptRunner
...
+            using (var connection = connectionBuilder(typeof(T)))
+            {
+                await connection.OpenAsync().ConfigureAwait(false);
+                using (var transaction = connection.BeginTransaction())
                 {
-                    await InstallTimeouts(scriptDirectory, connection, transaction, tablePrefix, sqlDialect).ConfigureAwait(false);
+                    await installAction(scriptDirectory, connection, transaction, tablePrefix, sqlDialect).ConfigureAwait(false);
                 }
-
-                transaction.Commit();
```
Comment (DavidBoike) [inline src/SqlPersistence/ScriptRunner.cs]: Should we worry that the installers won't run transactionally? Or, vs. the cost of trying to group things by connection builder and somehow execute common ones together in a transaction still?

**Comment by DavidBoike** [comment]
@SzymonPobiega Busted by APIApprover :wink:

File: src/SqlPersistence/Config/SqlPersistenceConfig_Connection.cs (hunk 1 above)
Comment (weralabaj) [reply src/SqlPersistence/Config/SqlPersistenceConfig_Connection.cs]: > also include the storage type it's trying to find.

👍

File: src/SqlPersistence/InstallerFeature.cs [hunk 3]
```diff
@@ -16,7 +16,11 @@ protected override void Setup(FeatureConfigurationContext context)
             return;
         }
 
-        settings.ConnectionBuilder = context.Settings.GetConnectionBuilder();
+        settings.ConnectionBuilder = storageType =>
```
Comment (weralabaj) [inline src/SqlPersistence/InstallerFeature.cs]: Could be `settings.ConnectionBuilder = storageType => context.Settings.GetConnectionBuilder(storageType)();` or you keep it that way for readability?

If that was the explicit intention then ignore comment.

File: src/SqlPersistence/ScriptRunner.cs (hunk 2 above)
Comment (weralabaj) [reply src/SqlPersistence/ScriptRunner.cs]: I think this is a bug? I don't see where those transactions are commited.

Comment (SzymonPobiega) [reply src/SqlPersistence/ScriptRunner.cs]: I don't think have a single transaction is necessary here as these installers are not destructive.

File: src/SqlPersistence/ScriptRunner.cs [hunk 4]
```diff
@@ -29,30 +40,37 @@ public static class ScriptRunner
...
 
-            using (var connection = await connectionBuilder.OpenConnection().ConfigureAwait(false))
-            using (var transaction = connection.BeginTransaction())
+            if (shouldInstallOutbox)
             {
-                if (shouldInstallOutbox)
-                {
-                    await InstallOutbox(scriptDirectory, connection, transaction, tablePrefix, sqlDialect).ConfigureAwait(false);
-                }
+                await Execute<StorageType.Outbox>(InstallOutbox, scriptDirectory, tablePrefix, sqlDialect, connectionBuilder).ConfigureAwait(false);
```
Comment (weralabaj) [inline src/SqlPersistence/ScriptRunner.cs]: I know this is used to avoid repetition, but I'd prefer to have that "execute" method inlined. It's really hard to follow this code :/

Or, the minimum would be to give this method a more descriptive name like `ExecuteScriptTransactionally` or something like that.

File: src/SqlPersistence/Config/SqlPersistenceConfig_Connection.cs [hunk 5]
```diff
@@ -19,13 +19,23 @@ public static void ConnectionBuilder(this PersistenceExtensions<SqlPersistence>
...
+            if (settings.TryGet($"SqlPersistence.ConnectionBuilder.{storageType.Name}", out Func<DbConnection> value))
             {
                 return value;
             }
-            throw new Exception("ConnectionBuilder must be defined.");
+            if (settings.TryGet("SqlPersistence.ConnectionBuilder", out value))
+            {
+                return value;
+            }
+            throw new Exception("The connection to the database must be specified using the ConnectionBuilder method.");
```
Comment (weralabaj) [inline src/SqlPersistence/Config/SqlPersistenceConfig_Connection.cs]: Suggestion:
` throw new Exception($"Couldn't find connection string for {storageType}. The connection to the database must be specified using the ConnectionBuilder method.");`

File: src/SqlPersistence/InstallerFeature.cs (hunk 3 above)
Comment (weralabaj) [reply src/SqlPersistence/InstallerFeature.cs]: ping

File: src/SqlPersistence/ScriptRunner.cs (hunk 4 above)
Comment (weralabaj) [reply src/SqlPersistence/ScriptRunner.cs]: ping

Comment (SzymonPobiega) [reply src/SqlPersistence/ScriptRunner.cs]: I am going with `ExecuteInSeparateConnection`.

**Comment by SzymonPobiega** [comment]
@weralabaj I applied your suggestions
